### PR TITLE
Use Streams in the WriteBinarySpecScript API

### PR DESCRIPTION
**Pull request by binji**
`WriteBinarySpecScript` writes a JSON file and its associated module
files (either `.wasm` or `.wat` files). The previous API wrote directly
to the filesystem, but it is more useful to allow the caller to specify
where to write the output instead.

The function now provides a callback `WriteBinarySpecStreamFactory`,
which takes a filename as parameter and returns a `Stream` to write to.

There is also a convenience function that will write the results to a
vector of filename/MemoryStream pairs.

### CODE REVIEW

**Comment by pepyakin** [comment]
Hello @sbc100 ! Could you please take a look at this?

File: src/binary-writer-spec.h [hunk 1]
```diff
@@ -17,21 +17,43 @@
...
-                             const WriteBinarySpecOptions*);
+typedef std::function<Stream*(string_view filename)>
+    WriteBinarySpecStreamFactory;
+
+Result WriteBinarySpecScript(Stream* json_stream,
+                             WriteBinarySpecStreamFactory module_stream_factory,
+                             Script*,
+                             string_view source_filename,
+                             string_view module_filename_noext,
+                             const WriteBinaryOptions*);
```
Comment (sbc100) [inline src/binary-writer-spec.h]: What are some args not named?  Is there some kind of rule here?

File: src/tools/wast2json.cc [hunk 2]
```diff
@@ -115,12 +116,23 @@ int ProgramMain(int argc, char** argv) {
...
+          StripExtension(s_outfile ? s_outfile : s_infile).to_string();
+      result = WriteBinarySpecScript(
+          &json_stream, script.get(), s_infile, module_filename_noext,
+          &s_write_binary_options, &module_streams, s_log_stream.get());
+
+      if (s_outfile) {
+        json_stream.WriteToFile(s_outfile);
+      }
+
+      std::for_each(module_streams.begin(), module_streams.end(),
```
Comment (sbc100) [inline src/tools/wast2json.cc]: Is this preferable to using a normal for loop?    Seems harder to visually parse and more verbose.

**Comment by sbc100** [comment]
`WriteBinarySpecStreamFactory` .. what is this Java project?

File: src/binary-writer-spec.h (hunk 1 above)
Comment (binji) [reply src/binary-writer-spec.h]: I often omit names that are redundant, kind of like Blink style. So you don't have `Script* script` or `const WriteBinaryOptions* options`. But it's not really anything formalized.

**Comment by binji** [comment]
> WriteBinarySpecStreamFactory .. what is this Java a project?

Ha! I didn't like it much either, do you have a better name?

**Pull request merged**